ssd_keras/ssd_box_encode_decode_utils.py

```diff
@@ -190,22 +211,58 @@ def greedy_nms(y_pred_decoded, iou_threshold=0.45, coords='minmax'):
-    function for per-class NMS in `decode_y()`.
-    '''
-    boxes_left = np.copy(predictions)
-    maxima = [] # This is where we store the boxes that make it through the non-maximum suppression
-    while boxes_left.shape[0] > 0: # While there are still boxes left to compare...
-        maximum_index = np.argmax(boxes_left[:,0]) # ...get the index of the next box with the highest confidence...
-        maximum_box = np.copy(boxes_left[maximum_index]) # ...copy that box and...
-        maxima.append(maximum_box) # ...append it to `maxima` because we'll definitely keep it
-        boxes_left = np.delete(boxes_left, maximum_index, axis=0) # Now remove the maximum box from `boxes_left`
-        if boxes_left.shape[0] == 0: break # If there are no boxes left after this step, break. Otherwise...
-        similarities = iou(boxes_left[:,1:], maximum_box[1:], coords=coords) # ...compare (IoU) the other left over boxes to the maximum box...
-        boxes_left = boxes_left[similarities <= iou_threshold] # ...so that we can remove the ones that overlap too much with the maximum box
-    return np.array(maxima)
+
+if NUMBA_INSTALLED:
+    @numba.jit(nopython=True)
+    def _greedy_nms(predictions, iou_threshold=0.45):
+        '''
+        The same greedy non-maximum suppression algorithm as above, but slightly modified for use as an internal
+        function for per-class NMS in `decode_y()`.
+        '''
+        boxes_left = np.copy(predictions)
+        # This is where we store the boxes that make it through the non-maximum suppression
+        maxima = []
+
+        # While there are still boxes left to compare...
+        while boxes_left.shape[0] > 0:
+            # ...get the index of the next box with the highest confidence...
+            maximum_index = np.argmax(boxes_left[:,0])
+            # ...copy that box and...
+            maximum_box = np.copy(boxes_left[maximum_index])
+            # ...append it to `maxima` because we'll definitely keep it
+            maxima.extend(list(maximum_box))
+            # Now remove the maximum box from `boxes_left`
+            #boxes_left = np.delete(boxes_left, maximum_index, axis=0)
+            boxes_left_new = np.empty((boxes_left.shape[0] - 1, 5), dtype=boxes_left.dtype)
+            boxes_left_new[:maximum_index, :] = boxes_left[:maximum_index, :]
+            boxes_left_new[maximum_index:, :] = boxes_left[maximum_index+1:, :]
+            boxes_left = boxes_left_new
+            # If there are no boxes left after this step, break.
+            if boxes_left.shape[0] == 0:
+                break
+            # Otherwise compare (IoU) the other left over boxes to the maximum box...
+            similarities = iou_specialized(boxes_left[:,1:], maximum_box[1:])
+            # ...so that we can remove the ones that overlap too much with the maximum box
+            boxes_left = boxes_left[similarities <= iou_threshold]
+        return np.array(maxima).reshape(-1, 5)
```

Comment:
Here the logic is also exactly the same, the main changes compared to the version below:

- np.delete replaced with manual recreation of new array without the specific value
- extend a flat list (and reshape at the end) instead of appending to the list, because numba cannot deal with list of arrays in nopython mode